CI Timing Statistics

kwryankrattiger opened the pull request:
Getting timing statistics for all of the packages installed in CI

Spack and Spack CI
* [x] Add timing logs to packages installed via a build cache
* [x] Rework timer to better handle nested timers (cache install phases extract and relocate)
* [x] Collect all install_times.json for upload
* [x] Include all timing logs in CI artifacts

Spack infrastructure
* [ ] Scrape timing logs and store data in database
* [ ] Create dashboard(s) for exploring timing data

CC: @tgamblin

kwryankrattiger committed:
Write timing information for installs from cache

kwryankrattiger committed:
CI: aggregate and upload install_times.json to artifacts

kwryankrattiger commented:
The result of this change is the packages installed from cache also dump a `install_times.json`and CI aggregates all of the `install_times.json` files for all packages into a single artifact for the job.

** Note ** two new fields are now present in the `install_times.json`, `name` and `cache`. One question I have to @tgamblin and others is whether or not it makes sense to also include the spec hash in this file, or what extra meta data should be included in the timer json output if any.

From cache:
[zlib-install_times.json.txt](https://github.com/spack/spack/files/11898748/zlib-install_times.json.txt)

From source:
[zstd-install_times.json.txt](https://github.com/spack/spack/files/11898749/zstd-install_times.json.txt)

Aggregated `install_times.json` files from all installed packages
[aggregated_install_times.json.txt](https://github.com/spack/spack/files/11898746/aggregated_install_times.json.txt)

kwryankrattiger commented on lib/spack/spack/installer.py:
```diff
@@ -86,6 +86,16 @@
 
+def _write_timer_json(pkg, timer, cache):
+    extra_attributes = {"name": pkg.name, "cache": cache}
+    try:
```
this try really shouldn't be required, but for some reason the `_install_from_cache` routine doesn't actually install anything in the `test_install_from_cache_ok` test.

haampie commented on share/spack/gitlab/cloud_pipelines/scripts/common/aggregate_package_logs.spack.py:
```diff
@@ -0,0 +1,44 @@
+"""
```
Use a `#!/usr/bin/env spack-python` shebang

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -123,23 +138,49 @@ def measure(self, name):
+        yield timer
+        timer.stop()
+
+    def subtimer(self, name):
```
This unused?

kwryankrattiger replied:
it is now, it is an alternative option to getting the timer from measure and what I first tried using for getting the subtimer, but it wasn't as pythonic. I am thinking about this tuple of events idea, I like that and it removes the subtle case where the timer just resets itself after start->stop->start which didn't feel great, and that might remove the need for this

haampie commented:
Just thinking out loud: what about an interface timer.start(identifier), timer.stop(identifier), with timer.measure(identifier), where all these operation just push an event tuple (time, start/stop, identifier)  to a list, and we do processing only afterwards by looping over the list? Then you can construct whatever type of format out of it, whether it's a tree, or a reduction on identifier, etc.

kwryankrattiger committed:
Flat event based timer variation

Event based timer allows for easily starting and stopping timers without
wiping sub-timer data. It also requires less branching logic when
tracking time.

The json output is non-hierarchical in this version and hierarchy is
less rigidly enforced between starting and stopping.

kwryankrattiger committed:
Add and write timers for top level install

kwryankrattiger committed:
CI: Don't change root directory for artifact generation

kwryankrattiger committed:
Update completion

kwryankrattiger committed:
remove unused subtimer api

kwryankrattiger committed:
Fix unit tests

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -2095,18 +2112,20 @@ def install(self) -> None:
             # other process may be hung).
             install_status.set_term_title("Acquiring lock for {0}".format(pkg.name))
             term_status.add(pkg_id)
+            t.start("aquire_lock")
```
```suggestion
            t.start("acquire_lock")
```

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -2095,18 +2112,20 @@ def install(self) -> None:
                 # the spec (or that process is hung).
                 ltype, lock = self._ensure_locked("read", pkg)
-
+            t.stop("aquire_lock")
```
```suggestion
            t.stop("acquire_lock")
```

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -2146,6 +2167,7 @@ def install(self) -> None:
                     # or uninstalled -- on the next pass.
                     self.installed.remove(pkg_id)
                     self._requeue_task(task, install_status)
+                t.stop("task")
```
seems like you really want to use a context manager here instead of repeating t.stop on every exit point

kwryankrattiger replied:
yes...but I also don't want to indent all of those lines either. Is there a way to c++ scope things without tabbing over all of the lines?

haampie replied:
nope and neither is there go like defer

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -2257,6 +2287,18 @@ def install(self) -> None:
         ]
 
+        t.stop()
+
+        phases = ["{}: {}.".format(p.capitalize(), _hms(t.duration(p))) for p in t.phases]
+        phases.append("Total: {}".format(_hms(t.duration())))
+        tty.msg("Installer time summary", "  ".join(phases))
```
```suggestion
```

General users will not understand why Spack would print timers about locks...

kwryankrattiger replied:
True. I  originally had it blocked by the summary flag. Does it make sense to only write it if the summary file is specified?

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -2257,6 +2287,18 @@ def install(self) -> None:
+            timer_summary_file = self.build_requests[0].install_args.get("timer_summary_file")
+            if timer_summary_file:
+                with open(timer_summary_file, "w") as fd:
+                    t.write_json(out=fd, extra_attributes={"name": "install_all"})
```
Shouldn't this also run on `^C`?

kwryankrattiger replied:
hmm...idk. do the timings still matter if the install was killed?

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/scripts/common/aggregate_package_logs.spack.py:
```diff
@@ -0,0 +1,38 @@
+    prefixes = [p for p in args.prefix.split(":") if os.path.exists(p)]
+
+    # Aggregate the install timers into a single json
+    data = []
```
Add information about runner here

haampie commented on lib/spack/spack/cmd/install.py:
```diff
@@ -87,6 +88,12 @@ def setup_parser(subparser):
+        "--timer-summary-file",
+        type=str,
+        default=None,
+        help="File to record the top level install timer summary",
```
```suggestion
        help=argparse.SUPPRESS,
```

let's not make it public

kwryankrattiger committed:
Suppress timing summary option

scottwittenburg commented:
@spackbot fix style

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/ci.yaml:
```diff
@@ -22,6 +22,10 @@ ci:
       script::
       - spack --color=always --backtrace ci rebuild --tests > >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_out.txt) 2> >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_err.txt >&2)
       after_script:
+      - - spack python ${CI_PROJECT_DIR}/share/spack/gitlab/cloud_pipelines/scripts/common/aggregate_package_logs.spack.py
```
This doesn't work for me, I think because any environment variables set in the `script` or `before_script` are not available in `after_script` [(docs)](https://docs.gitlab.com/ee/ci/variables/where_variables_can_be_used.html#execution-shell-environment).

kwryankrattiger replied:
All of those vars should be set in the `variables` section, so it should work. Only modifications from the other scripts aren't carried.

scottwittenburg replied:
I think the problematic environment variables are the ones set during the `. share/spack/setup-env.sh` in order to know what the `spack` command is.  But if you tested this and it worked for you, I'd be curious to know what's different about your setup.

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/ci.yaml:
```diff
@@ -22,6 +22,10 @@ ci:
+      - - spack python ${CI_PROJECT_DIR}/share/spack/gitlab/cloud_pipelines/scripts/common/aggregate_package_logs.spack.py
+          --prefix /home/software/spack:${CI_PROJECT_DIR}
+          --log install_times.json
+          ${SPACK_ARTIFACTS_ROOT}/install_times.json
```
Also, this artifact is ignored.  Could it go into `${SPACK_ARTIFACTS_ROOT}/logs/`?  We don't pick up as artifact just anything that appears in `$SPACK_ARTIFACTS_ROOT`, but only `logs/`, `reproduction/`, `tests/`, and `user_data/`.

kwryankrattiger replied:
yeah, I think I fixed that in a wip PR for testing but removed it during a rebase.

kwryankrattiger committed:
Save timers summaries to user_data artifacts

kwryankrattiger committed:
Remove completion from fish

kwryankrattiger committed:
Move spack python to script section

kwryankrattiger committed:
Write timer correctly for non-cache installs

kwryankrattiger committed:
Re-add hash to timer file

kwryankrattiger committed:
Fish completion updates

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -19,25 +19,26 @@
     @contextmanager
     def measure(self, name):
-        yield
+        yield NullTimer()
```
```suggestion
        yield self
```

haampie committed:
Fix null timer yield value

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -66,10 +67,12 @@ def __init__(self, now=time.time):
 
-        # _global is the overal timer since the instance was created
-        self._timers[global_timer_name] = Interval(self._now(), end=None)
+        self._events: list(TimerEvent) = []
```
```suggestion
        self._events: List[TimerEvent] = []
```

haampie committed:
fix type hints

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -123,23 +123,72 @@ def measure(self, name):
+        self._flatten()
+        return [k for k in self._timers.keys() if not k == global_timer_name]
+
+    def _flatten(self):
+        for event in self._events:
+            if event.running:
+                if event.label not in self._timer_stack:
+                    self._timer_stack.append(event.label)
+                # Only start the timer if it is on top of the stack
+                # restart doesn't work after a subtimer is started
+                if event.label == self._timer_stack[-1]:
+                    timer_path = "/".join(self._timer_stack[1:])
+                    tracker = self._timers.get(
+                        event.label, TimeTracker(0.0, event.time, 0, timer_path)
+                    )
+                    assert tracker.path == timer_path
+                    self._timers[event.label] = TimeTracker(
+                        tracker.total, event.time, tracker.count, tracker.path
+                    )
+            else:  # if not event.running:
+                if event.label in self._timer_stack:
+                    index = self._timer_stack.index(event.label)
+                    for label in self._timer_stack[index:]:
+                        tracker = self._timers[label]
+                        self._timers[label] = TimeTracker(
+                            tracker.total + (event.time - tracker.start),
+                            None,
+                            tracker.count + 1,
+                            tracker.path,
+                        )
+                    self._timer_stack = self._timer_stack[: max(0, index)]
+        # clear events
+        self._events = []
```
Can you follow up with a PR that makes this code actually readable? It's hard to see the intent, probably also for you after a couple months.

kwryankrattiger replied:
Like more comments, or re-written in a more pythonic way?

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -123,23 +123,72 @@ def measure(self, name):
+                            tracker.count + 1,
+                            tracker.path,
+                        )
+                    self._timer_stack = self._timer_stack[: max(0, index)]
```
(let's fix this in a follow-up PR) there's no need to use `max` here. In fact `del self._timer_stack[index:]` is what you want?

kwryankrattiger replied:
Basically I am trying to truncate all of the timer labels in the timer stack after and including "index"

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -123,23 +123,72 @@ def measure(self, name):
+                        tracker.total, event.time, tracker.count, tracker.path
+                    )
+            else:  # if not event.running:
+                if event.label in self._timer_stack:
```
What if this is not true? (let's fix this in a follow-up PR if necessary)

kwryankrattiger replied:
If it isn't true, you are stopping a timer that was never started, so it is just a noop. Is there something you think should be done in that case?

haampie commented on lib/spack/spack/util/timer.py:
```diff
@@ -123,23 +123,72 @@ def measure(self, name):
+                if event.label == self._timer_stack[-1]:
+                    timer_path = "/".join(self._timer_stack[1:])
+                    tracker = self._timers.get(
+                        event.label, TimeTracker(0.0, event.time, 0, timer_path)
```
(let's fix this in a follow-up PR) This always allocates a `TimeTracker(...)`, also when `self._timers[event.label]` exists. So, then it's better to use a standard if-else, or use defaultdict.

kwryankrattiger replied:
I was not aware of deafultdict, noted for the update later to improve the code clarity.

haampie commented:
Maybe I'm missing something, but (a) the flag `--timer-summary-file` is not set for Gitlab CI and (b) db lock acquire time only makes sense when installing dependencies in parallel, whereas this timer seems to be for a single process?

kwryankrattiger commented:
@haampie I talked about this with @scottwittenburg today and we agree this file is kind of useless. I originally wanted to do the timer aggregation with this flag, but since the parallelism is handled by Make and not Spack it is not so straightforward as to how to write the lock timer information. Instead of this flag, I added a timer to the `spack ci rebuild` sub-command to collect information on the unaccounted for overhead of runnering rebuild, ie. locks, makefile generation, etc. and will remove the `--time-summary-file`

kwryankrattiger committed:
Remove timer-summary-file option

kwryankrattiger committed:
Add "." in front of non-package timer name

scottwittenburg reviewed (approved):
I've been using this as the base for some branches I'm testing, [here](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines), and it seems to be providing the data we expect.  Unless there are objections, I vote to merge it.
